File: pulse.client/android/app/src/main/java/com/pixeldust/pulse/_REFERENCE_ONLY_health/HealthPlugin.kt
package com.pixeldust.pulse.health

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.util.Log
import androidx.activity.result.ActivityResult
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin
import androidx.health.connect.client.HealthConnectClient
import androidx.health.connect.client.PermissionController
import java.time.Instant
import java.time.Duration
import java.time.format.DateTimeParseException
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

@CapacitorPlugin(name = "Health")
class HealthPlugin : Plugin() {
    private val pluginVersion = "8.10.0"
    private val manager = HealthManager()
    private val pluginScope = CoroutineScope(
        SupervisorJob() + Dispatchers.Main.immediate +
            CoroutineExceptionHandler { _, throwable ->
                Log.e(TAG, "Unhandled coroutine failure", throwable)
            },
    )
    private val permissionContract = PermissionController.createRequestPermissionResultContract()

    // Store pending request data for callback
    private var pendingReadTypes: List<HealthDataType> = emptyList()
    private var pendingWriteTypes: List<HealthDataType> = emptyList()
    private var pendingIncludeWorkouts: Boolean = false
    private var pendingIncludeHistoryAccess: Boolean = false

    /**
     * Launch work on [pluginScope] and always settle [this] PluginCall on failure.
     * Without this, exceptions escaping a plain `launch` hit the thread's default
     * handler and kill the host app (e.g. Health Connect IPC rate limits).
     */
    private fun PluginCall.launchSafely(block: suspend PluginCall.() -> Unit) {
        val call = this
        pluginScope.launch {
            try {
                call.block()
            } catch (e: CancellationException) {
                throw e
            } catch (e: Throwable) {
                Log.e(TAG, "Health Connect call failed", e)
                val exception = e as? Exception ?: Exception(e)
                call.reject(e.message ?: "Health Connect call failed", null, exception)
            }
        }
    }

    override fun handleOnDestroy() {
        super.handleOnDestroy()
        pluginScope.cancel()
    }

    @PluginMethod
    fun isAvailable(call: PluginCall) {
        val status = HealthConnectClient.getSdkStatus(context)
        call.resolve(availabilityPayload(status))
    }

    @PluginMethod
    fun requestAuthorization(call: PluginCall) {
        val (readTypes, includeWorkouts) = try {
            parseTypeListWithWorkouts(call, "read")
        } catch (e: IllegalArgumentException) {
            call.reject(e.message, null, e)
            return
        }

        val writeTypes = try {
            parseTypeList(call, "write")
        } catch (e: IllegalArgumentException) {
            call.reject(e.message, null, e)
            return
        }

        val includeHistoryAccess = call.getBoolean("requestHistoryAccess") ?: false

        call.launchSafely {
            val client = getClientOrReject(this) ?: return@launchSafely
            // Only request the history permission when the provider actually supports it.
            // On an older-but-supported provider it can never be granted, which would leave
            // granted.containsAll(permissions) permanently false and reopen the permission
            // sheet on every call. The normal data scopes are still requested either way;
            // authorizationStatus reports historyAccessAvailable so callers can react.
            val requestHistoryAccess = includeHistoryAccess && manager.isHistoryAccessAvailable(client)
            val permissions = manager.permissionsFor(readTypes, writeTypes, includeWorkouts, requestHistoryAccess)

            if (permissions.isEmpty()) {
                val status = manager.authorizationStatus(client, readTypes, writeTypes, includeWorkouts, includeHistoryAccess)
                resolve(status)
                return@launchSafely
            }

            val granted = client.permissionController.getGrantedPermissions()
            if (granted.containsAll(permissions)) {
                val status = manager.authorizationStatus(client, readTypes, writeTypes, includeWorkouts, includeHistoryAccess)
                resolve(status)
                return@launchSafely
            }

            // Store types for callback
            pendingReadTypes = readTypes
            pendingWriteTypes = writeTypes
            pendingIncludeWorkouts = includeWorkouts
            pendingIncludeHistoryAccess = includeHistoryAccess

            // Create intent using the Health Connect permission contract
            val intent = permissionContract.createIntent(context, permissions)

            try {
                startActivityForResult(this, intent, "handlePermissionResult")
            } catch (e: Exception) {
                pendingReadTypes = emptyList()
                pendingWriteTypes = emptyList()
                reject("Failed to launch Health Connect permission request.", null, e)
            }
        }
    }

    @ActivityCallback
    private fun handlePermissionResult(call: PluginCall?, result: ActivityResult) {
        if (call == null) {
            return
        }

        val readTypes = pendingReadTypes
        val writeTypes = pendingWriteTypes
        val includeWorkouts = pendingIncludeWorkouts
        val includeHistoryAccess = pendingIncludeHistoryAccess
        pendingReadTypes = emptyList()
        pendingWriteTypes = emptyList()
        pendingIncludeWorkouts = false
        pendingIncludeHistoryAccess = false

        call.launchSafely {
            val client = getClientOrReject(this) ?: return@launchSafely
            val status = manager.authorizationStatus(client, readTypes, writeTypes, includeWorkouts, includeHistoryAccess)
            resolve(status)
        }
    }

    @PluginMethod
    fun checkAuthorization(call: PluginCall) {
        val (readTypes, includeWorkouts) = try {
            parseTypeListWithWorkouts(call, "read")
        } catch (e: IllegalArgumentException) {
            call.reject(e.message, null, e)
            return
        }

        val writeTypes = try {
            parseTypeList(call, "write")
        } catch (e: IllegalArgumentException) {
            call.reject(e.message, null, e)
            return
        }

        val includeHistoryAccess = call.getBoolean("requestHistoryAccess") ?: false

        call.launchSafely {
            val client = getClientOrReject(this) ?: return@launchSafely
            val status = manager.authorizationStatus(client, readTypes, writeTypes, includeWorkouts, includeHistoryAccess)
            resolve(status)
        }
    }

    @PluginMethod
    fun readSamples(call: PluginCall) {
        Log.d(TAG, "=== JS readSamples ENTERED ===")

        val identifier = call.getString("dataType")
        Log.d(TAG, "dataType=$identifier")

        if (identifier.isNullOrBlank()) {
            call.reject("dataType is required")
            return
        }

        val dataType = HealthDataType.from(identifier)
        Log.d(TAG, "resolved dataType=$dataType")

        if (dataType == null) {
            call.reject("Unsupported data type: $identifier")
            return
        }

        val limit = (call.getInt("limit") ?: DEFAULT_LIMIT).coerceAtLeast(0)
        val ascending = call.getBoolean("ascending") ?: false

        Log.d(TAG, "limit=$limit ascending=$ascending")

        val startInstant = try {
            manager.parseInstant(
                call.getString("startDate"),
                Instant.now().minus(DEFAULT_PAST_DURATION)
            )
        } catch (e: DateTimeParseException) {
            call.reject(e.message, null, e)
            return
        }

        val endInstant = try {
            manager.parseInstant(
                call.getString("endDate"),
                Instant.now()
            )
        } catch (e: DateTimeParseException) {
            call.reject(e.message, null, e)
            return
        }

        Log.d(TAG, "start=$startInstant end=$endInstant")

        if (endInstant.isBefore(startInstant)) {
            call.reject("endDate must be greater than or equal to startDate")
            return
        }

        call.launchSafely {
            Log.d(TAG, "=== ENTERED readSamples COROUTINE ===")

            val client = getClientOrReject(this) ?: return@launchSafely

            Log.d(TAG, "=== GOT HEALTH CONNECT CLIENT ===")

            try {
                Log.d(TAG, "=== CALLING HealthManager.readSamples ===")

                val samples = manager.readSamples(
                    client,
                    dataType,
                    startInstant,
                    endInstant,
                    limit,
                    ascending
                )

                Log.d(TAG, "=== HealthManager.readSamples RETURNED ===")

                val result = JSObject().apply {
                    put("samples", samples)
                }

                resolve(result)

                Log.d(TAG, "=== readSamples RESOLVED ===")
            } catch (e: Exception) {
                Log.e(TAG, "=== readSamples FAILED ===", e)

                reject(
                    e.message ?: "Failed to read samples.",
                    null,
                    e
                )
            }
        }
    }

    @PluginMethod
    fun saveSample(call: PluginCall) {
        val identifier = call.getString("dataType")
        if (identifier.isNullOrBlank()) {
            call.reject("dataType is required")
            return
        }

        val dataType = HealthDataType.from(identifier)
        if (dataType == null) {
            call.reject("Unsupported data type: $identifier")
            return
        }

        val value = call.getDouble("value")
        if (value == null) {
            call.reject("value is required")
            return
        }

        val unit = call.getString("unit")
        if (unit != null && unit != dataType.unit) {
            call.reject("Unsupported unit $unit for ${dataType.identifier}. Expected ${dataType.unit}.")
            return
        }

        val startInstant = try {
            manager.parseInstant(call.getString("startDate"), Instant.now())
        } catch (e: DateTimeParseException) {
            call.reject(e.message, null, e)
            return
        }

        val endInstant = try {
            manager.parseInstant(call.getString("endDate"), startInstant)
        } catch (e: DateTimeParseException) {
            call.reject(e.message, null, e)
            return
        }

        if (endInstant.isBefore(startInstant)) {
            call.reject("endDate must be greater than or equal to startDate")
            return
        }

        val metadataObj = call.getObject("metadata")
        val metadata = metadataObj?.let { obj ->
            val iterator = obj.keys()
            val map = mutableMapOf<String, String>()
            while (iterator.hasNext()) {
                val key = iterator.next()
                val rawValue = obj.opt(key)
                if (rawValue is String) {
                    map[key] = rawValue
                }
            }
            map.takeIf { it.isNotEmpty() }
        }
        
        val systolic = call.getDouble("systolic")
        val diastolic = call.getDouble("diastolic")
        val mindfulnessSessionType = call.getString("mindfulnessSessionType")

        call.launchSafely {
            val client = getClientOrReject(this) ?: return@launchSafely
            try {
                manager.saveSample(client, dataType, value, startInstant, endInstant, metadata, systolic, diastolic, mindfulnessSessionType)
                resolve()
            } catch (e: Exception) {
                reject(e.message ?: "Failed to save sample.", null, e)
            }
        }
    }

    private fun parseTypeList(call: PluginCall, key: String): List<HealthDataType> {
        val array = call.getArray(key) ?: JSArray()
        val result = mutableListOf<HealthDataType>()
        for (i in 0 until array.length()) {
            val identifier = array.optString(i, null) ?: continue
            val dataType = HealthDataType.from(identifier)
                ?: throw IllegalArgumentException("Unsupported data type: $identifier")
            result.add(dataType)
        }
        return result
    }

    private fun parseTypeListWithWorkouts(call: PluginCall, key: String): Pair<List<HealthDataType>, Boolean> {
        val array = call.getArray(key) ?: JSArray()
        val result = mutableListOf<HealthDataType>()
        var includeWorkouts = false
        for (i in 0 until array.length()) {
            val identifier = array.optString(i, null) ?: continue
            if (identifier == "workouts") {
                includeWorkouts = true
            } else {
                val dataType = HealthDataType.from(identifier)
                    ?: throw IllegalArgumentException("Unsupported data type: $identifier")
                result.add(dataType)
            }
        }
        return Pair(result, includeWorkouts)
    }

    private fun getClientOrReject(call: PluginCall): HealthConnectClient? {
        val status = HealthConnectClient.getSdkStatus(context)
        if (status != HealthConnectClient.SDK_AVAILABLE) {
            call.reject(availabilityReason(status))
            return null
        }
        return HealthConnectClient.getOrCreate(context)
    }

    private fun availabilityPayload(status: Int): JSObject {
        val payload = JSObject()
        payload.put("platform", "android")
        payload.put("available", status == HealthConnectClient.SDK_AVAILABLE)
        if (status != HealthConnectClient.SDK_AVAILABLE) {
            payload.put("reason", availabilityReason(status))
        }
        return payload
    }

    private fun availabilityReason(status: Int): String {
        return when (status) {
            HealthConnectClient.SDK_UNAVAILABLE_PROVIDER_UPDATE_REQUIRED -> "Health Connect needs an update."
            HealthConnectClient.SDK_UNAVAILABLE -> "Health Connect is unavailable on this device."
            else -> "Health Connect availability unknown."
        }
    }

    @PluginMethod
    fun getPluginVersion(call: PluginCall) {
        try {
            val ret = JSObject()
            ret.put("version", pluginVersion)
            call.resolve(ret)
        } catch (e: Exception) {
            call.reject("Could not get plugin version", e)
        }
    }

    @PluginMethod
    fun openHealthConnectSettings(call: PluginCall) {
        try {
            val intent = Intent(HealthConnectClient.ACTION_HEALTH_CONNECT_SETTINGS)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            context.startActivity(intent)
            call.resolve()
        } catch (e: Exception) {
            call.reject("Failed to open Health Connect settings", null, e)
        }
    }

    @PluginMethod
    fun showPrivacyPolicy(call: PluginCall) {
        try {
            val intent = Intent(context, PermissionsRationaleActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            context.startActivity(intent)
            call.resolve()
        } catch (e: Exception) {
            call.reject("Failed to show privacy policy", null, e)
        }
    }

    @PluginMethod
    fun queryWorkouts(call: PluginCall) {
        val workoutType = call.getString("workoutType")
        val limit = (call.getInt("limit") ?: DEFAULT_LIMIT).coerceAtLeast(0)
        val ascending = call.getBoolean("ascending") ?: false
        val anchor = call.getString("anchor")

        val startInstant = try {
            manager.parseInstant(call.getString("startDate"), Instant.now().minus(DEFAULT_PAST_DURATION))
        } catch (e: DateTimeParseException) {
            call.reject(e.message, null, e)
            return
        }

        val endInstant = try {
            manager.parseInstant(call.getString("endDate"), Instant.now())
        } catch (e: DateTimeParseException) {
            call.reject(e.message, null, e)
            return
        }

        if (endInstant.isBefore(startInstant)) {
            call.reject("endDate must be greater than or equal to startDate")
            return
        }

        call.launchSafely {
            val client = getClientOrReject(this) ?: return@launchSafely
            try {
                val result = manager.queryWorkouts(client, workoutType, startInstant, endInstant, limit, ascending, anchor)
                resolve(result)
            } catch (e: Exception) {
                reject(e.message ?: "Failed to query workouts.", null, e)
            }
        }
    }

    @PluginMethod
    fun queryAggregated(call: PluginCall) {
        val identifier = call.getString("dataType")
        if (identifier.isNullOrBlank()) {
            call.reject("dataType is required")
            return
        }

        val dataType = HealthDataType.from(identifier)
        if (dataType == null) {
            call.reject("Unsupported data type: $identifier")
            return
        }

        val bucket = call.getString("bucket") ?: "day"
        // `aggregation` may be a single string or an array of strings. Fall back to a
        // sensible per-data-type default when omitted.
        val aggregations: List<String> = call.getArray("aggregation")?.let { array ->
            try {
                (0 until array.length()).map { array.getString(it) }
            } catch (e: org.json.JSONException) {
                call.reject("aggregation array must contain only strings", null, e)
                return
            }
        } ?: call.getString("aggregation")?.let { listOf(it) } ?: listOf(
            when (dataType) {
                HealthDataType.STEPS,
                HealthDataType.DISTANCE,
                HealthDataType.CALORIES -> "sum"
                HealthDataType.HEART_RATE,
                HealthDataType.WEIGHT,
                HealthDataType.RESTING_HEART_RATE -> "average"
                else -> "sum"
            },
        )

        if (aggregations.isEmpty()) {
            call.reject("aggregation must not be empty")
            return
        }

        val startInstant = try {
            manager.parseInstant(call.getString("startDate"), Instant.now().minus(DEFAULT_PAST_DURATION))
        } catch (e: DateTimeParseException) {
            call.reject(e.message, null, e)
            return
        }

        val endInstant = try {
            manager.parseInstant(call.getString("endDate"), Instant.now())
        } catch (e: DateTimeParseException) {
            call.reject(e.message, null, e)
            return
        }

        if (endInstant.isBefore(startInstant)) {
            call.reject("endDate must be greater than or equal to startDate")
            return
        }

        call.launchSafely {
            val client = getClientOrReject(this) ?: return@launchSafely
            try {
                val result = manager.queryAggregated(client, dataType, startInstant, endInstant, bucket, aggregations)
                resolve(result)
            } catch (e: CancellationException) {
                throw e
            } catch (e: SecurityException) {
                Log.w(TAG, "Permission denied for aggregation: ${e.message}", e)
                reject(e.message ?: "Permission denied for aggregated data.", "permission-denied", e)
            } catch (e: IllegalArgumentException) {
                reject(e.message ?: "Unsupported aggregation.", null, e)
            } catch (e: Exception) {
                Log.w(TAG, "Aggregation failed: ${e.message}", e)
                reject(e.message ?: "Failed to query aggregated data.", "query-aggregated-failed", e)
            }
        }
    }

    companion object {
        private const val TAG = "HealthPlugin"
        private const val DEFAULT_LIMIT = 100
        private val DEFAULT_PAST_DURATION: Duration = Duration.ofDays(1)
    }
}

File: pulse.client/ios/Sources/HealthPlugin/HealthPlugin.swift
import Foundation
import Capacitor

@objc(HealthPlugin)
public class HealthPlugin: CAPPlugin, CAPBridgedPlugin {
    private let pluginVersion: String = "8.10.2"
    public let identifier = "HealthPlugin"
    public let jsName = "Health"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestAuthorization", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkAuthorization", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "readSamples", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "saveSample", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openHealthConnectSettings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "showPrivacyPolicy", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "queryWorkouts", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "queryAggregated", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = Health()

    @objc func isAvailable(_ call: CAPPluginCall) {
        call.resolve(implementation.availabilityPayload())
    }

    @objc func requestAuthorization(_ call: CAPPluginCall) {
        let read = (call.getArray("read") as? [String]) ?? []
        let write = (call.getArray("write") as? [String]) ?? []
        // `requestHistoryAccess` is Android-only (READ_HEALTH_DATA_HISTORY). HealthKit has no
        // equivalent permission and no 30-day read cap, so we intentionally ignore it here and
        // leave `historyAccessAuthorized` out of the returned status.

        implementation.requestAuthorization(readIdentifiers: read, writeIdentifiers: write) { result in
            DispatchQueue.main.async {
                switch result {
                case let .success(payload):
                    call.resolve(payload.toDictionary())
                case let .failure(error):
                    call.reject(error.localizedDescription, nil, error)
                }
            }
        }
    }

    @objc func checkAuthorization(_ call: CAPPluginCall) {
        let read = (call.getArray("read") as? [String]) ?? []
        let write = (call.getArray("write") as? [String]) ?? []
        // `requestHistoryAccess` is Android-only; intentionally ignored on iOS (see requestAuthorization).

        implementation.checkAuthorization(readIdentifiers: read, writeIdentifiers: write) { result in
            DispatchQueue.main.async {
                switch result {
                case let .success(payload):
                    call.resolve(payload.toDictionary())
                case let .failure(error):
                    call.reject(error.localizedDescription, nil, error)
                }
            }
        }
    }

    @objc func readSamples(_ call: CAPPluginCall) {
        guard let dataType = call.getString("dataType") else {
            call.reject("dataType is required")
            return
        }

        let startDate = call.getString("startDate")
        let endDate = call.getString("endDate")
        let limit = call.getInt("limit")
        let ascending = call.getBool("ascending") ?? false

        do {
            try implementation.readSamples(
                dataTypeIdentifier: dataType,
                startDateString: startDate,
                endDateString: endDate,
                limit: limit,
                ascending: ascending
            ) { result in
                DispatchQueue.main.async {
                    switch result {
                    case let .success(samples):
                        call.resolve(["samples": samples])
                    case let .failure(error):
                        call.reject(error.localizedDescription, nil, error)
                    }
                }
            }
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    @objc func saveSample(_ call: CAPPluginCall) {
        guard let dataType = call.getString("dataType") else {
            call.reject("dataType is required")
            return
        }

        guard let value = call.getDouble("value") else {
            call.reject("value is required")
            return
        }

        let unit = call.getString("unit")
        let startDate = call.getString("startDate")
        let endDate = call.getString("endDate")
        let metadataAny = call.getObject("metadata") as? [String: Any]
        let metadata = metadataAny?.reduce(into: [String: String]()) { result, entry in
            if let stringValue = entry.value as? String {
                result[entry.key] = stringValue
            }
        }
        
        let systolic = call.getDouble("systolic")
        let diastolic = call.getDouble("diastolic")

        do {
            try implementation.saveSample(
                dataTypeIdentifier: dataType,
                value: value,
                unitIdentifier: unit,
                startDateString: startDate,
                endDateString: endDate,
                metadata: metadata,
                systolic: systolic,
                diastolic: diastolic
            ) { result in
                DispatchQueue.main.async {
                    switch result {
                    case .success:
                        call.resolve()
                    case let .failure(error):
                        call.reject(error.localizedDescription, nil, error)
                    }
                }
            }
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": self.pluginVersion])
    }

    @objc func openHealthConnectSettings(_ call: CAPPluginCall) {
        // No-op on iOS - Health Connect is Android only
        call.resolve()
    }

    @objc func showPrivacyPolicy(_ call: CAPPluginCall) {
        // No-op on iOS - Health Connect privacy policy is Android only
        call.resolve()
    }

    @objc func queryWorkouts(_ call: CAPPluginCall) {
        let workoutType = call.getString("workoutType")
        let startDate = call.getString("startDate")
        let endDate = call.getString("endDate")
        let limit = call.getInt("limit")
        let ascending = call.getBool("ascending") ?? false
        let anchor = call.getString("anchor")

        implementation.queryWorkouts(
            workoutTypeString: workoutType,
            startDateString: startDate,
            endDateString: endDate,
            limit: limit,
            ascending: ascending,
            anchorString: anchor
        ) { result in
            DispatchQueue.main.async {
                switch result {
                case let .success(response):
                    call.resolve(response)
                case let .failure(error):
                    call.reject(error.localizedDescription, nil, error)
                }
            }
        }
    }

    @objc func queryAggregated(_ call: CAPPluginCall) {
        guard let dataType = call.getString("dataType") else {
            call.reject("dataType is required")
            return
        }

        let startDate = call.getString("startDate")
        let endDate = call.getString("endDate")
        let bucket = call.getString("bucket")

        // `aggregation` may be a single string or an array of strings. Reject a non-string
        // array explicitly rather than silently defaulting, matching Android's HealthPlugin.kt.
        var aggregations: [String] = []
        if let single = call.getString("aggregation") {
            aggregations = [single]
        } else if let rawArray = call.getArray("aggregation") {
            guard let array = rawArray as? [String] else {
                call.reject("aggregation array must contain only strings")
                return
            }
            aggregations = array
        }

        implementation.queryAggregated(
            dataTypeIdentifier: dataType,
            startDateString: startDate,
            endDateString: endDate,
            bucketString: bucket,
            aggregations: aggregations
        ) { result in
            DispatchQueue.main.async {
                switch result {
                case let .success(response):
                    call.resolve(response)
                case let .failure(error):
                    call.reject(error.localizedDescription, nil, error)
                }
            }
        }
    }

}
